### DISCUSSION TITLE
Incoming Clock to BPM

### DISCUSSION THREAD

**Discussion by ghost** [FortySevenEffects/arduino_midi_library#295] [Q&A] [answered]
how do i read an incoming midi clock and turn it to a BPM I can display?

**Comment by mink99** [comment]
Midi clock is basically 1/96 th notes.
So you need to measure one minute as exactly as possible, count the ticks, and do the maths. 
one bpm is a quarter note per minute.
Do not expect this to be exact, even tools like the beat step pro fail on this exercise…..

**Reply by franky47** [reply]
To catch the MIDI Clock messages, the simplest way is to use its associated callback:

```c++
void onClock()
{
  // Time maths goes here
}

void setup()
{
  MIDI.begin();
  MIDI.setHandleClock(onClock);
}

void loop()
{
  MIDI.read();
}

**Comment by ghost** [comment, marked as answer]
this prints bpm to serial monitor
thanks for the suggestions, solved.
`#include <MIDI.h>

byte midi_clock = 0xf8;
unsigned long int startTime;
int beatCount;
bool counting = false;
long int bpm;
long int noteTime;

MIDI_CREATE_INSTANCE(HardwareSerial, Serial1, MIDI);

void setup() {
  MIDI.begin(MIDI_CHANNEL_OMNI);
  Serial.begin(57600);
  Serial.println("MIDI Input Test");
}

void loop() {

  if (MIDI.read()) {               // Is there a MIDI message incoming ?
    byte type = MIDI.getType();
    if (type == midi_clock) {
      if ( not counting) { // first time you have seen a clock message in this batch
        startTime = millis();
        counting = true;
        beatCount = 1;
      }
      else {
        beatCount += 1;
        if (beatCount >= 96) {
          noteTime = millis() - startTime;
          counting = false;
          // you now have the time for one note so calculate the BPM
          bpm = 240000/noteTime;
          // now display the BPM
          Serial.println(bpm);
        }
      }
    }
  }
}`

**Reply by m-r-m-s** [reply]
Is there a way to print the bpm more quickly and/ or more dynamically? Using this method, if you are at a slower bpm, say 20-70 bpm, it takes a while for the bpm to print compared to 120 or higher.

**Reply by franky47** [reply]
There will always be a trade-off between stability and reactivity.

Either you only wait one period and get a noisy reading that jumps constantly (because of imprecisions in time measurement and jitter of the main loop), or you use a rolling average that will take time to stabilise.

Think of it like a "tap tempo" button, it usually takes a couple of taps to get the rhythm right.

That being said you don't have to wait for 96 Clock messages: computing the time between two subsequent Clocks, then multiplying it by 96 will give a good enough approximation. Feed that into a size 8 rolling average, and you should get better reactivity while keeping a somewhat stable reading.

**Reply by m-r-m-s** [reply]
> That being said you don't have to wait for 96 Clock messages: computing the time between two subsequent Clocks, then multiplying it by 96 will give a good enough approximation. Feed that into a size 8 rolling average, and you should get better reactivity while keeping a somewhat stable reading.

Ok, Thank you for this guidance. I think I understand the approach here in theory, but my coding and mathing chops are still pretty novice. Could you please provide some coding insight on adapting this example to your approach above? I realize I need to update the code to reflect 96 but I am not sure how to compute and get a rolling average? Is rolling average same as moving average? 

```
/*
 source: https://forum.arduino.cc/t/tempo-calculation-errors-when-sending-fast-midi-cc-messages-to-oled-display/866256/4
*/
unsigned long current_micros;
unsigned long previous_micros;
unsigned long beat_length;

unsigned int tempo = 120;                                     //round(60000000./beat_length (in micros)) 60000000us = 1s.
unsigned int previous_tempo;                               //Previous Tempo to store and detect tempo changes.

int clock_counter = 0;                                        //0 to 255, no more than 24 Midi Clock pulses by beat to calculate tempo.

bool updateClock() {
  if (clock_counter == 24) {                                  //If 24 loops, 24 Midi Clock pulses by beat, then resets counters. If not, +1 counter until 24.
    clock_counter = 1;                                        //Resets counter.
  }
  else {
    clock_counter++;                                          //+1 up to 24.
  }
  return clock_counter == 1;
}

void tempoDisplay() {
  char buffer_tempo_serial[8];                                    
  sprintf(buffer_tempo_serial, "%3d BPM", tempo);
  Serial.println(buffer_tempo_serial);                            //Prints tempo in BPM.
}

void onClock() {

  if (updateClock()) {
    
      current_micros = micros();
      beat_length = current_micros - previous_micros;  // is this where I would multiply by 96 and get a size 8 rolling average?
      previous_micros = current_micros;
      tempo = round(60000000. / beat_length);                   //tempo calculation. Using micros is more precise in higher tempos. round is not really needed.

      // tempoDisplay();                                        //Sends to tempoDisplay every 24 clock_counter.

      
      
    }

    if (tempo != previous_tempo && tempo < 1000) {              //Detects if tempo changes and tempo < 1000.
      tempoDisplay();                                           //Sends to tempoDisplay if tempo changes.
    }
    
    previous_tempo = tempo;

    
}

void setup () {
  MIDI.begin();
  MIDI.setHandleClock(onClock);
  Serial.begin(38400);
}

void loop () {
  MIDI.read();
}
```

**Reply by franky47** [reply]
> Is rolling average same as moving average?

Yes, here's how to implement one. It's advanced C++ with templates, but at the heart of it is a buffer of N elements (N being a power of two helps for optimisation of modulo operations), where we push a new item onto a moving head that wraps back once full, and compute the average over the whole buffer:

```c++
// Here we define some template arguments (compile-time variables of some sort)
// to bend the use of this object to suit our needs.
// Typenames allow configuring it to work on float, int, unsigned long long or whatever data type makes sense.
// Note that the SumType should always be bigger than the Sample type, to account for the accumulator when
// computing the average sum.
// We can also make the size (in bits) a template argument, to optimise things (compile-time calculations are optimised away).
template<unsigned SizeBits, typename Sample = unsigned long, typename SumType = unsigned long long>
struct RunningAverage
{
public:
  // Here we define some compile-time constants.
  // The size is a power of two of the order given by SizeBits (using a left-shift bit operation)
  // The mask is that size minus one, so it has SizeBits bits set to one, and the rest is zeros. We'll do fast modulo this way.
  static const unsigned sSize = 1 << SizeBits;
  static const unsigned sMask = sSize - 1;

public:
  // Constructor, initialising everything
  RunningAverage()
    : index(0)
    , usage(0)
  {
    memset(buffer, 0, sizeof(Sample) * sSize);
  }

public:
  // A reset call can be omitted since setup only runs once,
  // but it can be useful if the MIDI stream was cut off for example.
  inline void reset()
  {
    memset(buffer, 0, sizeof(Sample) * sSize);
    index = 0;
    usage = 0;
  }
  
  // Here we add our sample to the current index position,
  // Then move this index to the next slot, using a fast modulo by ANDing with the mask
  // Usage helps tracking at the beginning of filling up the buffer how many samples
  // we should compute the average upon, otherwise we get a long ramp-up until it fills up.
  inline void push(Sample input)
  {
    buffer[index] = input;
    index = (index + 1) & sMask;
    usage = min(usage + 1, sSize);
  }

  // Average code: sum all values in the buffer, then divide by the length.
  // Note that we always know the buffer will fill up from index zero when reset,
  // and that usage will equal the full capacity once ramped up.
  inline Sample average() const
  {
    if (usage == 0)
    {
      return 0;
    }
    SumType sum = 0;
    for (unsigned i = 0; i < usage; ++i)
    {
      sum += buffer[i];
    }
    return sum / (SumType)(usage);
  }

  // Not sure if necessary (copy-pasted from another project where I might have had a use for it).
  inline unsigned samples() const
  {
    return usage;
  }

  // This is where internal object data (members) are declared.
  // Every instance will carry with it a set of those members,
  // and methods called upon an instance will work on those (encapsulation).
private:
  Sample buffer[sSize];
  unsigned index;
  unsigned usage;
};
```

Now, there may have been some confusion between 96 and 24. Clock messages are sent at a rate of 24 messages per quarter note. Usually, tempo expressed in BPM corresponds to the number of quarter notes per minute (eg: at 120 bpm, every quarter note is separated by 0.5 second). The number 96 only makes sense when talking about how long is a bar of 4:4.

So every time you receive a Clock message, you'll measure the time elapsed since the last one. This is done by keeping the last received Clock time in a memory variable, and performing all the calculation on the onClock callback (no need to keep doing this over and over on the same data in the main loop):

```c++
#include <climits>

RunningAverage<3 /* 2^3 = 8 */> runningAverage;

unsigned long lastClockTime = 0;

void onClock()
{
  const unsigned long now = micros();

  if (lastClockTime == 0)
  {
    // This is the first Clock message we receive since the Arduino powered up,
    // so we initialise it and wait for the next one to measure an accurate interval.
    lastClockTime = now;
    return;
  }

  // Note that micros will overflow after 70 minutes,
  // so we need to take care of this case:
  const unsigned long elapsedTime = now > lastClockTime
    ? now - lastClockTime
    : ULONG_MAX - lastClockTime + now;
  
  // Push this into the rolling average
  rollingAverage.push(elapsedTime);
  
  // Compute the average elapsed time (in µs)
  const unsigned long averageElapsedTime = rollingAverage.average();

  // Convert to BPM
  const unsigned long bpm = 60000000 / averageElapsedTime;

  // Finally, update the memory variable with the current time
  lastClockTime = now;
}

void setup()
{
  runningAverage.reset();
  MIDI.setHandleClock(onClock);
  MIDI.begin();
}

void loop()
{
  MIDI.read();
}
```

This will handle the `micros` overflow every 70 minutes, and avoid large values when the first Clock message is received long after the Arduino was powered up. There will still be a ramp-up time when the tempo changes, but with a length 8 average, it will take a third of a quarter note to reach the correct value.

**Reply by m-r-m-s** [reply]
Ok wow - thank you!!! This is seriously pushing my limits of understanding C++ but in a good way. I _really_ appreciate it. I'll be studying this for a while. It is a great reference and motivator to keep learning! 

Ok, as I have been going through the code here, one thing I noticed for clarity is that the template uses`RunningAverage` but there are two references to `rollingAverage`  :  `rollingAverage.push(elapsedTime); `and `rollingAverage.average();` These need to be adjusted in order for everything to compile. 

Related, I have a question:
`RunningAverage<3 /* 2^3 = 8 */> runningAverage;` 
What is the purpose of the case change here? Is namespace or function reference of some kind?

**Reply by franky47** [reply]
Sorry if there are things that did not compile, I wrote the code out of the top of my head without taking the time to boot up the Arduino IDE to test it.

> What is the purpose of the case change here? Is namespace or function reference of some kind?

It's a coding convention: class/struct names often have PascalCase names (starting with an uppercase), and variables camelCase names. You're free to name them however you like of course.

What this line means is: create an instance of the struct `RunningAverage`, with template parameter 3 (and default values for the other two typenames, the inline comment added for clarity of why 3 becomes 8), into a variable named `runningAverage`. Then we work with that object under the variable name `runningAverage`, by calling methods (functions attached to the object) onto it, eg `runningAverage.push`.

**Reply by m-r-m-s** [reply]
No worries at all!  Thanks for clarifying the convention and how that line works. Much appreciated.

**Reply by franky47** [reply]
I have updated the average code block with comments that explains various parts. Enjoy, once you've learned templates it's hard to go back! :)

**Reply by m-r-m-s** [reply]
Yes, I can now see why = ) ! And thank you for the additional comments!

I think we need one small (but important) computation for the bpm, the clock rate of 24 messages per quarter note:
```
// Convert to BPM
  const unsigned long bpm = 60000000 / averageElapsedTime; 
```
Should be `const unsigned long bpm = 60000000 / 24 / averageElapsedTime;` correct?

**Reply by franky47** [reply]
Yes, of course! Kids, this is why you should not do maths after 10pm. 😅

**Reply by franky47** [reply]
Another fix: it should be `ULONG_MAX`, and not `ULLONG_MAX` (which is for an unsigned long long, which would cause a very weird behaviour at overflow). Those are defined in `<climits>` by the way, hence the include.

**Reply by m-r-m-s** [reply]
> Yes, of course! Kids, this is why you should not do maths after 10pm. 😅

 I am in full support of no maths after 10pm 😅 and thanks for the fix 🙌

**Reply by ghost** [reply]
way beyond my understanding all this!
can you post a full sketch that compiles?

**Reply by franky47** [reply]
This should do it (untested on a hardware setup as I don't have boards at hand at the moment).

Note: `<climit>` does not seem to be available in the AVR libc headers, but `<limits.h>` (the C counterpart) works as a replacement.

```c++
#include <MIDI.h>
#include <limits.h>

MIDI_CREATE_INSTANCE(HardwareSerial, Serial1, MIDI);

// --

// Here we define some template arguments (compile-time variables of some sort)
// to bend the use of this object to suit our needs.
// Typenames allow configuring it to work on float, int, unsigned long long or whatever data type makes sense.
// Note that the SumType should always be bigger than the Sample type, to account for the accumulator when
// computing the average sum.
// We can also make the size (in bits) a template argument, to optimise things (compile-time calculations are optimised away).
template<unsigned SizeBits, typename Sample = unsigned long, typename SumType = unsigned long long>
struct RunningAverage
{
public:
  // Here we define some compile-time constants.
  // The size is a power of two of the order given by SizeBits (using a left-shift bit operation)
  // The mask is that size minus one, so it has SizeBits bits set to one, and the rest is zeros. We'll do fast modulo this way.
  static const unsigned sSize = 1 << SizeBits;
  static const unsigned sMask = sSize - 1;

public:
  // Constructor, initialising everything
  RunningAverage()
    : index(0)
    , usage(0)
  {
    memset(buffer, 0, sizeof(Sample) * sSize);
  }

public:
  // A reset call can be omitted since setup only runs once,
  // but it can be useful if the MIDI stream was cut off for example.
  inline void reset()
  {
    memset(buffer, 0, sizeof(Sample) * sSize);
    index = 0;
    usage = 0;
  }
  
  // Here we add our sample to the current index position,
  // Then move this index to the next slot, using a fast modulo by ANDing with the mask
  // Usage helps tracking at the beginning of filling up the buffer how many samples
  // we should compute the average upon, otherwise we get a long ramp-up until it fills up.
  inline void push(Sample input)
  {
    buffer[index] = input;
    index = (index + 1) & sMask;
    usage = min(usage + 1, sSize);
  }

  // Average code: sum all values in the buffer, then divide by the length.
  // Note that we always know the buffer will fill up from index zero when reset,
  // and that usage will equal the full capacity once ramped up.
  inline Sample average() const
  {
    if (usage == 0)
    {
      return 0;
    }
    SumType sum = 0;
    for (unsigned i = 0; i < usage; ++i)
    {
      sum += buffer[i];
    }
    return sum / (SumType)(usage);
  }

  // This is where internal object data (members) are declared.
  // Every instance will carry with it a set of those members,
  // and methods called upon an instance will work on those (encapsulation).
private:
  Sample buffer[sSize];
  unsigned index;
  unsigned usage;
};

// 2^3 = 8, at 24ppq it should ramp-up in a 1/3rd of a quarter note.
RunningAverage<3> runningAverage;

unsigned long lastClockTime = 0;
unsigned long lastBPM = 0;

// --

void onClock()
{
  const unsigned long now = micros();

  if (lastClockTime == 0)
  {
    // This is the first Clock message we receive since the Arduino powered up,
    // so we initialise it and wait for the next one to measure an accurate interval.
    lastClockTime = now;
    return;
  }

  // Note that micros will overflow after 70 minutes,
  // so we need to take care of this case:
  const unsigned long elapsedTime = now > lastClockTime
    ? now - lastClockTime
    : ULONG_MAX - lastClockTime + now;
  
  // Push this into the rolling average
  runningAverage.push(elapsedTime);
  
  // Compute the average elapsed time (in µs)
  const unsigned long averageElapsedTime = runningAverage.average();

  // Convert to BPM
  const unsigned long bpm = 60000000 / (averageElapsedTime * 24);

  // Print only when there are changes
  if (lastBPM != bpm)
  {
    Serial.print(bpm);
    Serial.println(" BPM");
    lastBPM = bpm;  
  }

  // Finally, update the memory variable with the current time
  lastClockTime = now;
}

// --

void setup()
{
  runningAverage.reset();
  MIDI.begin(MIDI_CHANNEL_OMNI);
  MIDI.setHandleClock(onClock);
  Serial.begin(115200);
  while(!Serial);
  Serial.println("Arduino ready");
}

void loop()
{
  MIDI.read();
}
```

**Reply by ghost** [reply]
Awesome! Thank you! ❤️

**Reply by m-r-m-s** [reply]
Again, thank you for all the detailed explanation and refactoring the previous code! Compiled and tested on Teensy LC and DIN MIDI circuit. With C++ templates, is the coding convention to have them inline with the code as shown here or can they also be in a separate file? For example: could I name the template `runningaverage.h` and then use `#include<runningaverage.h>` ?

**Reply by franky47** [reply]
They definitely can be in a separate file (it's a good practice), but it has to be a header, not compiled on its own in a .cpp file.  This is because the implementation has to be available at the point where it's used for the template to be "specialised" (ie: conformed to the provided template arguments), it's not something that can be linked like the rest of statically known types (eg: a class or function without template arguments, where it's fine to only include the declaration for usage and compile the implementation separately).

**Reply by franky47** [reply]
Note: I found the repo where I copy-pasted the running average from, it's my project to sync a Line6 DL4 to a MIDI clock. If you want to learn more about C++ and templates, you can have a peek:

https://github.com/franky47/dl4-midi-sync

**Reply by m-r-m-s** [reply]
Oh this is a great project! I have a DL4 so I may give this a try at some point! And very helpful as a learning guide too.

**Reply by franky47** [reply]
There's a bit of soldering to do to connect things together, I'll try and dig out a blog post draft I started writing about it.

**Reply by franky47** [reply]
Here's the blog post, which explains how the whole thing came together:
https://github.com/franky47/dl4-midi-sync/blob/main/blog-post.md

**Reply by m-r-m-s** [reply]
> Here's the blog post, which explains how the whole thing came together: https://github.com/franky47/dl4-midi-sync/blob/main/blog-post.md

Fantastic - thanks again 🙌!

**Comment by mink99** [comment]
Please keep in mind that Serial.println is incredibly slow. Even if you are on a mega and use Serial1 for midi , if you are measuring permanently, the println will tamper the measuring. We are not on a multiprocessor machine here.

**Comment by mink99** [comment]
Additionally there is always jitter on any device distributing midi clock. If there is a PC using a cheap usb midi interface, it is rather unreliable to put out any measuring. (And just forget about blutooth midi)
